doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md

```diff
@@ -0,0 +1,691 @@
+
+    - Number of VRFs - 1024
+    - Number of VLANs - 4096
+    - Number of DHCP Servers per VLAN - 32
+
+-  **R13:** This proposed DHCP relay agent will need to support all the functionality that has been added over the years in the community through various patches. The complete backward compatibitlity with ISC DHCP is an aspirational goal.
+
+
+# Design Considerations
+
+## Co-existence with ISC-DHCP Code
+To facilitate a smooth transition for users migrating from ISC-DHCP to the new SONiC-DHCPv4-Relay, both implementations will coexist within the SONiC codebase for a period of time. A new configuration flag, has_sonic_dhcpv4_relay, will be introduced in the dhcp_relay feature within the config-db. This flag will allow users to select either the ISC-DHCP or the SONiC-DHCPv4-Relay implementation, with the default set to the existing ISC-DHCP. Once the new design has been validated as a functional superset of ISC-DHCP, both the feature flag and the ISC-DHCP implementation will be deprecated.
+
+## Alignment with DHCPv6 Relay
+The new DHCPv4 relay design will aim to mirror the design structure of the existing DHCPv6 relay as closely as possible, ensuring consistency and ease of integration. However, the code for the DHCPv4 relay will be maintained in a new repository: https://github.com/sonic-net/sonic-dhcpv4-relay.
+
+## Interop with Port-Based DHCP Server
+In the current ISC-DHCP design, a daemon process named `dhcprelayd` operates within the dhcp_relay container. If `dhcp_server` feature is enabled, this daemon is responsible for monitoring the `dhcpv4_server` configuration and managing the lifecycle of the `dhcrelay` process, including actions such as starting, stopping, and restarting it.
+
+The new DHCPv4 Relay design aims to continue this integration with the port-based DHCPv4 server feature by not requiring the need for explicit relay configuration. Instead of depending on an external daemon, the new design will rely on a DHCPv4 CfgMgr thread (described later in this document), which will continuously monitor both the ConfigDb and StateDb for any port-based DHCP server configurations. This thread will dynamically configure the relay functionality as needed, ensuring that changes in server configurations are seamlessly integrated without extra configuration.
+
+## DCHP Monitor
+The existing DHCP monitor will be enhanced in-order to support monitoring of DHCPv4 packets being handled by the new implementation.
+
+## Dual-Tor Support
+In a dual-TOR (Top-of-Rack) architecture, it's possible for DHCP request packets from a client to reach the server through one TOR, while the server's response might arrive at the peer TOR. If the peer TOR has its link in standby mode, it won't be able to relay the response back to the originating client. This issue can be addressed by configuring the DHCPv4 relay with the link-selection and source-interface options.
```

Comment:
Could you please add detail of dualtor? i.e. What is the value of giaddr of tx packets. How dual tor proceed rx packets? Seems you plan to use vlan-id in circuit id, if customizing option 82 is feasible, can we fallback to giaddr or yiddr if hostname is not set?
@lolyu I think dst ip of reply packets sent by DHCP server would be giaddr in DHCP packets. Is it acceptable to set giaddr as vlan ip in dualtor? Then DHCP server would send DHCP reply packets with dst ip set as vlan ip.

Reply:
Giaddr of the the packets forwarded by the relay will be dependent on the DHCP relay configuration. If the link-selection and source-interface options are set as described below, giaddr will be set to the Loopback IP of the relay.
For the received packet, if option 82 doesn't have the necessary information to identify the downstream VLAN interface , then giaddr will be used to identify it. We can add yiddr also as a secondary fallback option if it's needed for dual-tor operation. Please let us know.